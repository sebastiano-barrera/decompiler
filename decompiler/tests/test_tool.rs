//! Tests based on "test_tool"
//!
//! These are pretty weak, but they work on real code. They only check that the
//! decompiler doesn't panic, and they write the decompiler's output as text to
//! a version-controlled file, just so that it's somewhere easy to diff and hard
//! to forget updating.

use include_dir::{include_dir, Dir};
use insta::assert_snapshot;

use std::{
    path::Path,
    sync::{Mutex, MutexGuard, OnceLock},
};

static DATA_DIR: Dir<'_> = include_dir!("$CARGO_MANIFEST_DIR/test-data/");

macro_rules! case {
    ($case:ident, $exe:expr, $func_name:ident) => {
        #[test_log::test]
        #[allow(non_snake_case)]
        fn $case() {
            let function_name = stringify!($func_name);
            let out = $exe.process_function(function_name);
            crate::assert_snapshot!(out);
        }
    };
}

macro_rules! tests_in_binary {
    ($group:ident, $exe_path:expr; $($funcs:ident),* $(,)?) => {
        mod $group {
            use crate::Exe;
            static EXE: Exe = Exe::new($exe_path);

            $(case!($funcs, EXE, $funcs);)*
        }
    };
}

struct Exe {
    once_lock: OnceLock<Mutex<decompiler::Executable<'static>>>,
    path: &'static str,
}
impl Exe {
    const fn new(path: &'static str) -> Self {
        Exe {
            once_lock: OnceLock::new(),
            path,
        }
    }

    fn get_or_init(&self) -> MutexGuard<'_, decompiler::Executable<'static>> {
        let mutex = self.once_lock.get_or_init(|| {
            let rel_path = Path::new(self.path);
            let raw = DATA_DIR.get_file(rel_path).unwrap().contents();
            let exe = decompiler::Executable::parse(raw).unwrap();
            Mutex::new(exe)
        });
        mutex.lock().unwrap()
    }

    fn process_function(&self, function_name: &str) -> String {
        use std::fmt::Write;

        let mut exe = self.get_or_init();

        let df = exe
            .decompile_function(function_name)
            .expect("decompiling function");

        let mut log_buf = String::new();

        if let Some(mil) = df.mil() {
            writeln!(log_buf, " --- mil").unwrap();
            writeln!(log_buf, "{:?}\n", mil).unwrap();
        }

        if let Some(ssa) = df.ssa_pre_xform() {
            writeln!(log_buf, " --- ssa pre-xform").unwrap();
            writeln!(log_buf, "{:?}\n", ssa).unwrap();
        }

        if let Some(ssa) = df.ssa() {
            writeln!(log_buf, " --- cfg").unwrap();
            let cfg = ssa.cfg();
            writeln!(log_buf, "  entry: {:?}", cfg.direct().entry_bid()).unwrap();
            for bid in cfg.block_ids() {
                let regs: Vec<_> = ssa.block_regs(bid).collect();
                writeln!(
                    log_buf,
                    "  {:?} -> {:?} {:?}",
                    bid,
                    cfg.block_cont(bid),
                    regs
                )
                .unwrap();
            }
            write!(log_buf, "  domtree:\n    ").unwrap();

            let mut pp_buf = decompiler::pp::FmtAsIoUTF8(&mut log_buf);
            let pp = &mut decompiler::pp::PrettyPrinter::start(&mut pp_buf);
            cfg.dom_tree().dump(pp).unwrap();
            writeln!(log_buf).unwrap();

            writeln!(log_buf, " --- ssa").unwrap();
            writeln!(log_buf, "{:?}\n", ssa).unwrap();

            writeln!(log_buf, " --- ast").unwrap();
            let mut ast = decompiler::Ast::new(&ssa, exe.types());
            let mut pp_buf = decompiler::pp::FmtAsIoUTF8(&mut log_buf);
            let pp = &mut decompiler::pp::PrettyPrinter::start(&mut pp_buf);
            ast.pretty_print(pp).unwrap();
        }

        log_buf
    }
}

tests_in_binary!(
    exe_composite_type, "ty/test_composite_type.so";
    list_len
);
tests_in_binary!(exe_capstone, "integration/sample_capstone"; main);
tests_in_binary!(exe_matrix, "integration/sample_matrix"; sum_matrix);
tests_in_binary!(
    redis_server, "integration/redis-server";
    ctl_lookup,
    listNext,
    listTypePush,
    geoArrayCreate,
    // regression: infinite loop
    processAnnotations,
    redisReconnect,
    redisConnectWithOptions,
);
