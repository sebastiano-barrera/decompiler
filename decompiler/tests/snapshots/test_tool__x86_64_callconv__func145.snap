---
source: decompiler/tests/test_tool.rs
expression: out
---
func145 0x4013f1+4 (file 0x13f1)
          4013f1: 0f 28 c1                movaps xmm0,xmm1
          4013f4: c3                      ret

function type: "func145" func (arg0 *void <TypeID(1v1)>,
                               arg1 *void <TypeID(1v1)>,
                               arg2 *void <TypeID(1v1)>,
                               arg3 *void <TypeID(1v1)>,
                               arg4 *void <TypeID(1v1)>,
                               arg5 small <TypeID(785v1)>,
                               arg6 small_xmms <TypeID(792v1)>) double <TypeID(784v1)>
0 warnings.


mil program = 
program  76 instrs
    0 r12: TypeID(2v1) <- Ancestral(AncestralName("stack_bottom"))
    1 r2: TypeID(2v1) <- Ancestral(AncestralName("CF"))
    2 r3: TypeID(2v1) <- Ancestral(AncestralName("PF"))
    3 r4: TypeID(2v1) <- Ancestral(AncestralName("AF"))
    4 r5: TypeID(2v1) <- Ancestral(AncestralName("ZF"))
    5 r6: TypeID(2v1) <- Ancestral(AncestralName("SF"))
    6 r7: TypeID(2v1) <- Ancestral(AncestralName("TF"))
    7 r8: TypeID(2v1) <- Ancestral(AncestralName("IF"))
    8 r9: TypeID(2v1) <- Ancestral(AncestralName("DF"))
    9 r10: TypeID(2v1) <- Ancestral(AncestralName("OF"))
   10 r11: TypeID(2v1) <- Ancestral(AncestralName("RBP"))
   11 r12: TypeID(2v1) <- Ancestral(AncestralName("RSP"))
   12 r13: TypeID(2v1) <- Ancestral(AncestralName("RIP"))
   13 r14: TypeID(2v1) <- Ancestral(AncestralName("RDI"))
   14 r15: TypeID(2v1) <- Ancestral(AncestralName("RSI"))
   15 r16: TypeID(2v1) <- Ancestral(AncestralName("RAX"))
   16 r17: TypeID(2v1) <- Ancestral(AncestralName("RBX"))
   17 r18: TypeID(2v1) <- Ancestral(AncestralName("RCX"))
   18 r19: TypeID(2v1) <- Ancestral(AncestralName("RDX"))
   19 r20: TypeID(2v1) <- Ancestral(AncestralName("R8"))
   20 r21: TypeID(2v1) <- Ancestral(AncestralName("R9"))
   21 r22: TypeID(2v1) <- Ancestral(AncestralName("R10"))
   22 r23: TypeID(2v1) <- Ancestral(AncestralName("R11"))
   23 r24: TypeID(2v1) <- Ancestral(AncestralName("R12"))
   24 r25: TypeID(2v1) <- Ancestral(AncestralName("R13"))
   25 r26: TypeID(2v1) <- Ancestral(AncestralName("R14"))
   26 r27: TypeID(2v1) <- Ancestral(AncestralName("R15"))
   27 r28: TypeID(2v1) <- Ancestral(AncestralName("ZMM0"))
   28 r29: TypeID(2v1) <- Ancestral(AncestralName("ZMM1"))
   29 r30: TypeID(2v1) <- Ancestral(AncestralName("ZMM2"))
   30 r31: TypeID(2v1) <- Ancestral(AncestralName("ZMM3"))
   31 r32: TypeID(2v1) <- Ancestral(AncestralName("ZMM4"))
   32 r33: TypeID(2v1) <- Ancestral(AncestralName("ZMM5"))
   33 r34: TypeID(2v1) <- Ancestral(AncestralName("ZMM6"))
   34 r35: TypeID(2v1) <- Ancestral(AncestralName("ZMM7"))
   35 r36: TypeID(2v1) <- Ancestral(AncestralName("ZMM8"))
   36 r37: TypeID(2v1) <- Ancestral(AncestralName("ZMM9"))
   37 r38: TypeID(2v1) <- Ancestral(AncestralName("ZMM10"))
   38 r39: TypeID(2v1) <- Ancestral(AncestralName("ZMM11"))
   39 r40: TypeID(2v1) <- Ancestral(AncestralName("ZMM12"))
   40 r41: TypeID(2v1) <- Ancestral(AncestralName("ZMM13"))
   41 r42: TypeID(2v1) <- Ancestral(AncestralName("ZMM14"))
   42 r43: TypeID(2v1) <- Ancestral(AncestralName("ZMM15"))
   43 r45: TypeID(2v1) <- Ancestral(AncestralName("arg0"))
   44 r15: TypeID(2v1) <- Get(r45)
   45 r46: TypeID(2v1) <- Ancestral(AncestralName("arg1"))
   46 r19: TypeID(2v1) <- Get(r46)
   47 r47: TypeID(2v1) <- Ancestral(AncestralName("arg2"))
   48 r18: TypeID(2v1) <- Get(r47)
   49 r48: TypeID(2v1) <- Ancestral(AncestralName("arg3"))
   50 r20: TypeID(2v1) <- Get(r48)
   51 r49: TypeID(2v1) <- Ancestral(AncestralName("arg4"))
   52 r21: TypeID(2v1) <- Get(r49)
   53 r50: TypeID(2v1) <- Ancestral(AncestralName("arg5"))
   54 r52: TypeID(2v1) <- Part { src: r50, offset: 0, size: 8 }
   55 r51: TypeID(2v1) <- ArithK(Add, r12, 8)
   56 r51: TypeID(2v1) <- StoreMem { addr: r51, value: r52 }
   57 r52: TypeID(2v1) <- Part { src: r50, offset: 8, size: 8 }
   58 r51: TypeID(2v1) <- ArithK(Add, r12, 16)
   59 r51: TypeID(2v1) <- StoreMem { addr: r51, value: r52 }
   60 r53: TypeID(2v1) <- Ancestral(AncestralName("arg6"))
   61 r54: TypeID(2v1) <- Part { src: r53, offset: 0, size: 8 }
   62 r55: TypeID(2v1) <- Part { src: r28, offset: 8, size: 56 }
   63 r28: TypeID(2v1) <- Concat { lo: r54, hi: r55 }
   64 r56: TypeID(2v1) <- Part { src: r53, offset: 8, size: 8 }
   65 r57: TypeID(2v1) <- Part { src: r29, offset: 8, size: 56 }
   66 r29: TypeID(2v1) <- Concat { lo: r56, hi: r57 }
0x4013f1:
   67 r46: TypeID(2v1) <- Part { src: r29, offset: 0, size: 16 }
   68 r47: TypeID(2v1) <- Part { src: r28, offset: 16, size: 48 }
   69 r28: TypeID(2v1) <- Concat { lo: r46, hi: r47 }
0x4013f4:
   70 r45: TypeID(2v1) <- Void
   71 r46: TypeID(2v1) <- Part { src: r28, offset: 0, size: 8 }
   72 r45: TypeID(2v1) <- Concat { lo: r45, hi: r46 }
   73 r45: TypeID(2v1) <- Part { src: r45, offset: 0, size: 8 }
   74 r47: TypeID(2v1) <- SetReturnValue(r45)
   75 r47: TypeID(2v1) <- Control(Ret)


ssa pre-xform:
ssa program  76 instrs
.B0:    ;;  → Always(Return).
  (  2)  r11: ? <- Ancestral(AncestralName("RSP"))
         r27: ? <- Ancestral(AncestralName("ZMM0"))
         r28: ? <- Ancestral(AncestralName("ZMM1"))
  (  2)  r53: ? <- Ancestral(AncestralName("arg5"))
         r54: ? <- Part { src: r53, offset: 0, size: 8 }
         r55: ? <- ArithK(Add, r11, 8)
         r56: ? <- StoreMem { addr: r55, value: r54 }
         r57: ? <- Part { src: r53, offset: 8, size: 8 }
         r58: ? <- ArithK(Add, r11, 16)
         r59: ? <- StoreMem { addr: r58, value: r57 }
  (  2)  r60: ? <- Ancestral(AncestralName("arg6"))
         r61: ? <- Part { src: r60, offset: 0, size: 8 }
         r62: ? <- Part { src: r27, offset: 8, size: 56 }
         r63: ? <- Concat { lo: r61, hi: r62 }
         r64: ? <- Part { src: r60, offset: 8, size: 8 }
         r65: ? <- Part { src: r28, offset: 8, size: 56 }
         r66: ? <- Concat { lo: r64, hi: r65 }
         r67: ? <- Part { src: r66, offset: 0, size: 16 }
         r68: ? <- Part { src: r63, offset: 16, size: 48 }
         r69: ? <- Concat { lo: r67, hi: r68 }
         r71: ? <- Part { src: r69, offset: 0, size: 8 }
         r72: ? <- Concat { lo: r70, hi: r71 }
         r73: ? <- Part { src: r72, offset: 0, size: 8 }
         r74: ? <- SetReturnValue(r73)


cfg:
  entry: BlockID(0)
  BlockID(0) -> Always(Return) [r11, r27, r28, r53, r54, r55, r56, r57, r58, r59, r60, r61, r62, r63, r64, r65, r66, r67, r68, r69, r70, r71, r72, r73, r74]
  domtree:
    BlockID(0)

ssa post-xform:
ssa program  76 instrs
.B0:    ;;  → Always(Return).
  (  2)  r11: ? <- Ancestral(AncestralName("RSP"))
  (  2)  r53: ? <- Ancestral(AncestralName("arg5"))
         r54: ? <- Part { src: r53, offset: 0, size: 8 }
         r55: ? <- ArithK(Add, r11, 8)
         r56: ? <- StoreMem { addr: r55, value: r54 }
         r57: ? <- Part { src: r53, offset: 8, size: 8 }
         r58: ? <- ArithK(Add, r11, 16)
         r59: ? <- StoreMem { addr: r58, value: r57 }
         r60: ? <- Ancestral(AncestralName("arg6"))
         r73: ? <- Part { src: r60, offset: 8, size: 8 }
         r74: ? <- SetReturnValue(r73)



T0: {
  [pre:RSP + 8]:* := pre:arg5[8 .. 0];
  [pre:RSP + 16]:* := pre:arg5[16 .. 8];
  ;
  return pre:arg6[16 .. 8]
}
